include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -0,0 +1,250 @@
+    using sec_aligned_t = typename traits_t::sec_aligned_t; //!< The aligned sequence type for the second range.
+
+public:
+
+    //!\brief The result type when building the aligned sequences.
+    struct result_type
+    {
+        matrix_coordinate begin; //!< The coordinate where the trimmed first and second sequence begin.
+        matrix_coordinate end; //!< The coordinate where the trimmed first and second sequence end.
+        std::pair<fst_aligned_t, sec_aligned_t> alignment; //!< The alignment over the trimmed sequences.
+    };
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr aligned_sequence_builder() = default; //!< Defaulted.
+    constexpr aligned_sequence_builder(aligned_sequence_builder const &) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder(aligned_sequence_builder &&) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder & operator=(aligned_sequence_builder const &) = default; //!< Defaulted.
+    constexpr aligned_sequence_builder & operator=(aligned_sequence_builder &&) = default; //!< Defaulted.
+    ~aligned_sequence_builder() = default; //!< Defaulted.
+
+    /*\brief Construction from the underlying sequences
+     * \param[in] fst_rng The first range to build the aligned sequence for.
+     * \param[in] sec_rng The second range to build the aligned sequence for.
+     */
+    constexpr aligned_sequence_builder(fst_rng_t fst_rng, sec_rng_t sec_rng) :
+        fst_rng{view::all(std::forward<fst_rng_t>(fst_rng))},
+        sec_rng{view::all(std::forward<sec_rng_t>(sec_rng))}
+    {}
+    //!\}
+
+    /*!\brief Builds the aligned sequences from the given trace path
+     * \tparam trace_path_t The type of the trace path; must model std::ranges::InputRange and
+     *                      std::Same<value_type_t<trace_path_t>, seqan::detail::trace_directions> must evaluate to
+     *                      `true`.
+     * \param[in] trace_path The trace path.
+     * \returns seqan3::detail::aligned_sequence_builder::result_type with the built alignment.
+     *
+     * \details
+     *
+     * From the given trace path this function builds the aligned sequences for the first and the second
+     * target sequence. The return type seqan3::detail::aligned_sequence_builder::result_type is an aggregate
+     * type containing the begin and end coordinates with the respective sequence positions and the respective alignment
+     * over the sliced sequences.
+     */
+    template <std::ranges::InputRange trace_path_t>
+    result_type build_from(trace_path_t && trace_path)
+    {
+        static_assert(std::Same<value_type_t<trace_path_t>, trace_directions>,
+                      "The value type of the trace path must be seqan3::detail::trace_directions");
+
+        auto trace_it = std::ranges::begin(trace_path);
+        matrix_coordinate path_end = trace_it.coordinate();
+
+        std::vector<std::pair<trace_directions, size_t>> traces;
```

Comment:
why do you fill this vector instead of directly going over the path and adding gaps? I am not familiar with the trace path but I imagine it to be something like `{up, diag, left, left, diag}`.  So why iterating over it twice and filling a vector instead of going along the trace until it changes and inserting the respective gaps directly? (It should still be done in the separate function of course.)

Reply:
because you need to parse the trace from right to left (from sink to source in matrix view). 
Therefor, whenever you add something in that order you need to update all the virtual positions to the right. Depeding on the underlying DS, this can be quadratic in runtime, whereas this implementation is linear by first getting all trace segments and then applying them from right to left.

Reply:
But the current DS is the gap_decorator in most cases correct (Because why should you use a std::list insead of a std::vector)? And for the gap_decorator, it does not matter performance-wise if you add gaps at the front or back (red-black-tree). So now you add extra overhead that is not needed right now and makes the code slower because you expect that in the future this might be different? What happened to incremental refactoring? And maybe we should always use the most efficient way possible given a certain DS

Reply:
Ok, I just looked into the implementation and it seems that insert is always linear. Also note, that the documentation is wrong about this. So no matter which version it is always quadratic 😭.
So once the insert is implemented efficiently to find the location in the set the current version is as fast as if I insert from right to left. But note that if the requirements for a gap decorator are not fulfilled a std::vector<gapped> is used. In this case inserting from left to right is also better than inserting from right to left.

Reply:
**EDIT**: I just got confused about your comments. In general the current version is always at least as good as the insertion from left to right into the underlying DS. 
a: Either we store the Virtual positions directly on the elements (as it is right now). In this case you always need to scan all nodes right of it to update the VP. So this is always quadratic, since I have n inserts and n updates per insert, while in my case the algorithm is linear.
b: If you would store only an offset on the inner nodes to be added for the right subtree you need n*log(n) time for n elements to insert. The same for my version. linear scan and then log(n) per node to insert.